include/divergence_block_chain.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include "communicator.h"
+#include "recorders.h"
+#include <mutex>
+#include <thread>
+#include <vector>
+
+namespace celerity::detail {
+/**
+ *  @brief This class is a wrapper around a 1D vector that allows us to access it as a 2D array.
+ *
+ *  It is used to send the task hashes to other nodes using MPI while keeping the code simple and readable.
+ */
+template <typename T>
+struct vector_2d {
+  public:
+	vector_2d(size_t width, size_t height) : m_data(width * height), m_width(width){};
+
+	const T& operator[](id<2> ij) const {
+		assert(ij[0] * m_width + ij[1] < m_data.size());
+		return m_data[ij[0] * m_width + ij[1]];
+	}
+
+	T* data() { return m_data.data(); }
+
+  private:
+	std::vector<T> m_data;
+	const size_t m_width;
+};
```

Comment:
```suggestion
/**
 * Stores the hashes of tasks for each node.
 * The data is stored densely so it can easily be exchanged through MPI collective operations.
 */
struct per_node_task_hashes {
  public:
	per_node_task_hashes(const size_t max_hash_count, const size_t num_nodes) : m_data(max_hash_count * num_nodes), m_max_hash_count(max_hash_count){};
	const task_hash& operator()(const node_id nid, const size_t i) const { return m_data.at(nid * m_max_hash_count + i); }
	task_hash* data() { return m_data.data(); }

  private:
	std::vector<task_hash> m_data;
	size_t m_max_hash_count;
};
```

No need to go overly generic on this if it's only being used in one place. For clarity I'd suggest something like this instead.

Reply:
Also move into `divergence_block_chain_detail` namespace!